File: examples/typescript/onboard.ts
/**
 * # OpenConstruct — Agent Onboarding (TypeScript)
 *
 * Registers an agent with the OpenConstruct coordinator and maintains
 * a heartbeat loop. Uses the official `@openconstruct/sdk` package.
 *
 * ## Usage
 *   npm install @openconstruct/sdk
 *   npx ts-node onboard.ts -- --coordinator ws://localhost:9142
 */

import {
  CoordinatorClient,
  AgentIdentity,
  Capability,
  type Session,
} from "@openconstruct/sdk";

const DEFAULT_COORDINATOR = "ws://localhost:9142";
const DEFAULT_NAME = "typescript-agent";
const HEARTBEAT_INTERVAL_MS = 30_000;

// ── Parse CLI args minimally ────────────────────────────────────────
function parseArgs(): { coordinator: string; name: string } {
  const args = process.argv.slice(2);
  let coordinator = DEFAULT_COORDINATOR;
  let name = DEFAULT_NAME;

  for (let i = 0; i < args.length; i++) {
    if (args[i] === "--coordinator" && args[i + 1]) {
      coordinator = args[++i];
    } else if (args[i] === "--name" && args[i + 1]) {
      name = args[++i];
    }
  }

  return { coordinator, name };
}

// ── Heartbeat loop ──────────────────────────────────────────────────
async function heartbeatLoop(session: Session): Promise<void> {
  const interval = setInterval(async () => {
    try {
      const status = await session.heartbeat();
      console.log(`[heartbeat] ok — ${JSON.stringify(status)}`);
    } catch (err) {
      console.warn(`[heartbeat] failed: ${err}`);
    }
  }, HEARTBEAT_INTERVAL_MS);

  // Clean up on SIGINT
  process.on("SIGINT", () => {
    clearInterval(interval);
    session.disconnect();
    console.log("disconnected — bye");
    process.exit(0);
  });
}

// ── Main ────────────────────────────────────────────────────────────
async function main(): Promise<void> {
  const { coordinator, name } = parseArgs();

  // Connect
  const client = new CoordinatorClient(coordinator);
  await client.connect();
  console.log(`connected to ${coordinator}`);

  // Declare identity
  const identity: AgentIdentity = {
    name,
    capabilities: [Capability.Sense, Capability.Act],
    metadata: {
      language: "typescript",
      example: "onboard",
    },
  };

  // Onboard
  const session = await client.onboard(identity);
  console.log(
    `onboarded as ${session.agentId} — assigned to room ${session.assignedRoom}`
  );

  // Heartbeat loop
  await heartbeatLoop(session);
}

main().catch((err) => {
  console.error("fatal:", err);
  process.exit(1);
});

File: examples/python/onboard.py
#!/usr/bin/env python3
"""
# OpenConstruct — Agent Onboarding (Python)

Thin-client example that registers an agent with the OpenConstruct
coordinator and enters a heartbeat loop using the official Python SDK.

## Usage
    pip install openconstruct
    python onboard.py --coordinator ws://localhost:9142
"""

from __future__ import annotations

import argparse
import logging
import signal
import sys
import time

from openconstruct import AgentIdentity, Capability, CoordinatorClient

logging.basicConfig(
    level=logging.INFO,
    format="%(asctime)s [%(levelname)s] %(name)s: %(message)s",
)
log = logging.getLogger("onboard")


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(description="OpenConstruct Python onboarding example")
    parser.add_argument(
        "--coordinator",
        default="ws://localhost:9142",
        help="Coordinator WebSocket URL",
    )
    parser.add_argument("--name", default="python-agent", help="Agent name")
    return parser.parse_args()


def main() -> None:
    args = parse_args()

    # ── Connect to the coordinator ───────────────────────────────────
    client = CoordinatorClient.connect(args.coordinator)
    log.info("connected to %s", args.coordinator)

    # ── Declare identity and capabilities ────────────────────────────
    identity = AgentIdentity(
        name=args.name,
        capabilities=[Capability.SENSE, Capability.ACT],
        metadata={
            "language": "python",
            "example": "onboard",
        },
    )

    # ── Onboard ──────────────────────────────────────────────────────
    session = client.onboard(identity)
    log.info(
        "onboarded as %s — assigned to room %s",
        session.agent_id,
        session.assigned_room,
    )

    # ── Graceful shutdown on SIGINT ──────────────────────────────────
    running = True

    def _shutdown(signum: int, _frame: object) -> None:
        nonlocal running
        log.info("received signal %d, shutting down", signum)
        running = False

    signal.signal(signal.SIGINT, _shutdown)
    signal.signal(signal.SIGTERM, _shutdown)

    # ── Heartbeat loop ───────────────────────────────────────────────
    while running:
        try:
            status = session.heartbeat()
            log.info("heartbeat ok — %s", status)
        except Exception as exc:
            log.warning("heartbeat failed: %s", exc)

        for _ in range(30):  # 30 × 1s = 30s sleep, interruptible
            if not running:
                break
            time.sleep(1)

    # ── Clean disconnect ─────────────────────────────────────────────
    session.disconnect()
    log.info("disconnected — bye")
    sys.exit(0)


if __name__ == "__main__":
    main()
